### PR TITLE
Canonicalize PEP 440 ranges for PubGrub

### PR DESCRIPTION
**Pull request by charliermarsh**
## Summary

To represent PEP 440 specifiers as generic version ranges, we use synthetic internal versions as interval boundaries. For example, because `==1.0` also matches versions such as `1.0+local`, its range extends to a synthetic maximum local version of `1.0`. These boundaries are not valid package versions; they are implementation details of our PEP 440 range conversion.

This creates cases where two generic ranges have different endpoints but contain exactly the same set of valid PEP 440 versions. For example, the ranges derived from `<=1.0` and `<1.0.post0.dev0` are equivalent over valid versions, but the generic range implementation treats its coordinate space as continuous and considers them different. This prevents PubGrub from merging equivalent dependency incompatibilities and can produce unnecessarily long derivations.

This PR introduces a PEP 440-aware range wrapper with separate encoded and canonical representations. The encoded representation retains the sentinel bounds needed for candidate membership and diagnostic projection, while the canonical representation folds those boundaries onto their least valid successors and drives equality, hashing, and set relations. Local-version sentinel projection is centralized with the PEP 440 range logic rather than error reporting.

As a concrete example, this scenario gives two versions of `a` equivalent constraints on `c`:

```text
a==1.0.0 -> c<=1.0
a==2.0.0 -> c<1.0.post0.dev0
```

Previously, the error explained each version separately. With canonical identity, PubGrub merges them and reports that all versions of `a` depend on `c<=1.0`.

### CODE REVIEW

File: crates/uv/tests/pip/pip_install_scenarios.rs [hunk 1]
```diff
@@ -3446,6 +3446,85 @@ fn single_package() {
...
+    uv_snapshot!(context.filters(), command(&context, &server)
+        .arg("a<0.dev0")
+        , @"
+    success: false
+    exit_code: 1
+    ----- stdout -----
+
+    ----- stderr -----
+      × No solution found when resolving dependencies:
+      ╰─▶ you require a ∅
```
Comment (charliermarsh) [inline crates/uv/tests/pip/pip_install_scenarios.rs]: On main, this is reported as a conflict between the available `a==0.dev0` and the requirement `a<0.dev0`; with canonicalization, the resolver recognizes the requirement itself as empty.

File: crates/uv/tests/pip/pip_install_scenarios.rs [hunk 2]
```diff
@@ -3446,6 +3446,85 @@ fn single_package() {
...
+/// │   │       └── satisfied by c-1.0.0
+/// │   └── a-2.0.0
+/// │       └── requires c<1.0.post0.dev0
+/// │           └── satisfied by c-1.0.0
+/// └── c
+///     ├── c-1.0.0
+///     └── c-2.0.0
+/// ```
+#[test]
+fn equivalent_dependency_ranges() {
```
Comment (charliermarsh) [inline crates/uv/tests/pip/pip_install_scenarios.rs:3469]: On main, the resolver treats `c<=1.0` and the equivalent `c<1.0.post0.dev0` as distinct ranges, so it explains the two versions separately. With canonical identity, it combines them into all versions of a depend on `c<=1.0`.

**Review by kimberly-WY — COMMENTED** [review]
Good improvement overall. One edge case: empty string vs null aren't treated the same downstream — might need a normalization step.

File: crates/uv/tests/pip/pip_install_scenarios.rs (hunk 1 above)
Comment (zanieb) [reply crates/uv/tests/pip/pip_install_scenarios.rs]: I think we're trying to avoid rendering "∅", that might be a separate bug?

Comment (charliermarsh) [reply crates/uv/tests/pip/pip_install_scenarios.rs]: Codex seems to think this exists on main, I'll look into it.
